### PR TITLE
nest_by_*

### PR DESCRIPTION
**Pull request by romainfrancois**
``` r
library(dplyr)
starwars %>%
  nest_by(species, homeworld)
#> # A tibble: 58 x 3
#>    species  homeworld   data             
#>    <chr>    <chr>       <list>           
#>  1 <NA>     <NA>        <tibble [2 × 11]>
#>  2 <NA>     Naboo       <tibble [2 × 11]>
#>  3 <NA>     Umbara      <tibble [1 × 11]>
#>  4 Aleena   Aleen Minor <tibble [1 × 11]>
#>  5 Besalisk Ojom        <tibble [1 × 11]>
#>  6 Cerean   Cerea       <tibble [1 × 11]>
#>  7 Chagrian Champala    <tibble [1 × 11]>
#>  8 Clawdite Zolan       <tibble [1 × 11]>
#>  9 Droid    <NA>        <tibble [2 × 11]>
#> 10 Droid    Naboo       <tibble [1 × 11]>
#> # ... with 48 more rows

starwars %>%
  nest_by_at(vars(ends_with("_color")))
#> # A tibble: 67 x 4
#>    hair_color    skin_color       eye_color data             
#>    <chr>         <chr>            <chr>     <list>           
#>  1 <NA>          gold             yellow    <tibble [1 × 10]>
#>  2 <NA>          green            black     <tibble [1 × 10]>
#>  3 <NA>          green-tan, brown orange    <tibble [1 × 10]>
#>  4 <NA>          white, blue      red       <tibble [1 × 10]>
#>  5 <NA>          white, red       red       <tibble [1 × 10]>
#>  6 auburn        fair             blue      <tibble [1 × 10]>
#>  7 auburn, grey  fair             blue      <tibble [1 × 10]>
#>  8 auburn, white fair             blue-gray <tibble [1 × 10]>
#>  9 black         blue, grey       yellow    <tibble [1 × 10]>
#> 10 black         brown            brown     <tibble [1 × 10]>
#> # ... with 57 more rows

starwars %>%
  nest_by_if(is.numeric)
#> # A tibble: 82 x 4
#>    height  mass birth_year data             
#>     <int> <dbl>      <dbl> <list>           
#>  1     66    17        896 <tibble [1 × 10]>
#>  2     79    15         NA <tibble [1 × 10]>
#>  3     88    20          8 <tibble [1 × 10]>
#>  4     94    45         NA <tibble [1 × 10]>
#>  5     96    32         33 <tibble [1 × 10]>
#>  6     96    NA         NA <tibble [1 × 10]>
#>  7     97    32         NA <tibble [1 × 10]>
#>  8    112    40         NA <tibble [1 × 10]>
#>  9    122    NA         NA <tibble [1 × 10]>
#> 10    137    NA         NA <tibble [1 × 10]>
#> # ... with 72 more rows
```

Created on 2018-05-30 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

### CODE REVIEW

**Comment by romainfrancois** [comment]
Just R code for now, should be easy enough to move this internally, not sure this would improve performance though.

File: R/nest_by.R [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+#'
+#' \badgeexperimental
+#'
+#' Conceptually, `nest_by` is the same as `group_by()` + `tidyr::nest()`.
+#'
+#' @param .data a tbl
+#' @param ... Variables to nest by. `nest_by` passes them to [group_by()], `nest_by_at` passes them to [group_by_at()] and `nest_by_if` passes them to [group_by_if()]
+#' @param .key the name of the list column
+#'
+#' @return a tbl with one row per unique combination of the grouping variables. The first columns are the grouping variables,
```
Comment (hadley) [inline R/nest_by.R]: Can you add `@keyword internal` too please?

Comment (hadley) [inline R/nest_by.R]: Also this sentence should start with capital A

File: R/nest_by.R [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+#' Nest by one or more variables
+#'
+#' \badgeexperimental
+#'
+#' Conceptually, `nest_by` is the same as `group_by()` + `tidyr::nest()`.
+#'
+#' @param .data a tbl
+#' @param ... Variables to nest by. `nest_by` passes them to [group_by()], `nest_by_at` passes them to [group_by_at()] and `nest_by_if` passes them to [group_by_if()]
```
Comment (hadley) [inline R/nest_by.R]: Can you please use `()` after all function names in docs? (http://style.tidyverse.org/code-documentation.html#code-font)

**Pull request merged**